rust/main/agents/relayer/src/msg/metadata/message_builder.rs

```diff
@@ -0,0 +1,427 @@
+#![allow(clippy::blocks_in_conditions)] // TODO: `rustc` 1.80.1 clippy issue
+#![allow(clippy::unnecessary_get_then_check)] // TODO: `rustc` 1.80.1 clippy issue
+
+use std::{fmt::Debug, sync::Arc};
+
+use async_trait::async_trait;
+use eyre::{Context, Result};
+use hyperlane_core::{HyperlaneMessage, InterchainSecurityModule, ModuleType, H256};
+
+use tokio::sync::Mutex;
+use tracing::instrument;
+
+use crate::msg::metadata::base_builder::BaseMetadataBuilderTrait;
+
+use super::{
+    aggregation::AggregationIsmMetadataBuilder,
+    base::{IsmWithMetadataAndType, MetadataBuildError},
+    ccip_read::CcipReadIsmMetadataBuilder,
+    metadata_builder::MessageMetadataBuildParams,
+    multisig::{MerkleRootMultisigMetadataBuilder, MessageIdMultisigMetadataBuilder},
+    routing::RoutingIsmMetadataBuilder,
+    Metadata, MetadataBuilder,
+};
+
+/// Builds metadata for a message.
+#[derive(Debug, Clone)]
+pub struct MessageMetadataBuilder {
+    pub base: Arc<Box<dyn BaseMetadataBuilderTrait>>,
+    pub app_context: Option<String>,
+    /// current ISM depth.
+    /// ISMs can be structured recursively. We keep track of the depth
+    /// of the recursion to avoid infinite loops.
+    pub depth: u32,
+    /// current ISM count
+    pub ism_count: Arc<Mutex<u32>>,
```

Comment:
We update `depth` and `ism_count` in the same place. I wonder why we need `Arc<Mutex<u32>>` for `ism_count` and we don't need it for `depth`?

Reply:
For AggregationISM, we spawn tokio threads to run the ISM metadata building in parallel.
These threads don't really need to share the depth because the depths are independent. But they need to share the count because one ISM might take up a lot of ISMs.